### ISSUE TITLE
Zero-inflation instability remediation — master roadmap (A → E → D/B → C/F)

### ISSUE THREAD

**Issue by Polichinel** [views-platform/views-hydranet#42]
## Overview

This is the **master tracking issue** for remediating the autoregressive divergence caused by zero-inflation handling in the HydraNet model. It coordinates 6 sub-issues across 4 phases, from immediate stability fix to strategic architecture exploration.

### Root cause

The sweep isolation experiment (`reports/sweep_isolation_plan.md`) identified `hurdle_threshold=0.0` as the primary instability source. The hard hurdle mask zeros out regression loss for ~95% of grid cells (y=0), starving the regression head of gradient. During 36-step autoregressive inference, unconstrained zero-cell predictions compound and diverge.

### The fix is not one thing

The literature survey (10 papers reviewed, see `reports/paths_forward.md`) reveals the problem has multiple layers:

| Layer | Problem | Fix |
|-------|---------|-----|
| **Loss function** | Gradient starvation from hurdle mask | Tobit censored likelihood (Path A) |
| **Training procedure** | Exposure bias from teacher forcing | GTF / scheduled sampling (Path E) |
| **Tail calibration** | Normal assumption underpredicts extremes | GP tail component or qs99 (Path D) |
| **Selection bias** | Regression head trained on censored sample | Heckman correction (Path B) |
| **Generative model** | Censoring vs mixture for zero-inflation | ZINB likelihood (Path C) |
| **Architecture** | Autoregressive error compounding | Direct multi-horizon (Path F) |

---

## Recommended Sequence

```
                    ┌─────────────────────────────────┐
                    │  PHASE 1: Fix the root cause    │
                    │                                 │
                    │  #36 Path A: Tobit loss          │
                    │  (replace hurdle mask)           │
                    │  Effort: 1-2 days               │
                    │  Risk: Low-Medium               │
                    └────────────┬────────────────────┘
                                 │
                          A passes?
                         ╱         ╲
                       Yes          No → investigate σ sensitivity,
                        │               try reparametrized variant
                        ▼
                    ┌─────────────────────────────────┐
                    │  PHASE 2: Stabilize inference   │
                    │                                 │
                    │  #37 Path E: Scheduled sampling  │
                    │  (then GTF if needed)            │
                    │  Effort: 2-3 days               │
                    │  Risk: Low-Medium               │
                    └────────────┬────────────────────┘
                                 │
                          E validated?
                         ╱         ╲
                       Yes          No → escalate to GTF;
                        │               if still poor, consider Path F
                        ▼
          ┌──────────────────────────────────────────────┐
          │  PHASE 3: Refine calibration (parallel)      │
          │                                              │
          │  #38 Path D          │  #39 Path B            │
          │  Tail calibration    │  Heckman correction    │
          │  (GP or keep qs99)   │  (couple cls+reg heads)│
          │  Effort: 2-3 days   │  Effort: 3-5 days     │
          │  Risk: Low          │  Risk: Medium          │
          │                                              │
          │  Diagnostic-gated: only pursue if metrics    │
          │  show tail underestimation (D) or selection  │
          │  bias (B) after Paths A+E are running.       │
          └──────────────┬───────────────────────────────┘
                         │
                  Quality sufficient?
                 ╱                  ╲
               Yes                   No
                │                     │
                ▼                     ▼
           ┌─────────┐    ┌──────────────────────────────┐
           │  Done.   │    │  PHASE 4: Strategic redesign  │
           │  Ship it.│    │                               │
           └─────────┘    │  #40 Path C: ZINB likelihood   │
                          │  Effort: 2-4 weeks             │
                          │                                │
                          │  #41 Path F: Direct MH         │
                          │  Effort: 10-16 weeks           │
                          │                                │
                          │  These combine naturally into   │
                          │  a DynAttn-style architecture   │
                          │  (Iacus et al. 2025, validated  │
                          │  on our exact VIEWS/PRIO-Grid   │
                          │  data).                        │
                          └────────────────────────────────┘
```

---

## Issue Index

| Phase | Issue | Path | Title | Depends on | Effort |
|-------|-------|------|-------|-----------|--------|
| **1** | #36 | **A** | Tobit censored-normal likelihood | None | 1-2 days |
| **2** | #37 | **E** | Generalized Teacher Forcing | #36 | 2-3 days |
| **3a** | #38 | **D** | Tail-aware extension (GP / qs99) | #36 | 2-3 days |
| **3b** | #39 | **B** | Heckman selection coupling | #36 | 3-5 days |
| **4a** | #40 | **C** | ZINB likelihood | #36, #37, #38, #39 outcomes | 2-4 weeks |
| **4b** | #41 | **F** | Direct multi-horizon forecasting | #36, #37, #38, #39 outcomes | 10-16 weeks |

---

### Gate 1: After Path A (#36)

**Question:** Does Tobit loss restore training convergence and autoregressive stability?

- **If yes:** Proceed to Path E. The loss function fix was sufficient for stability; now optimize inference quality.
- **If no:** Investigate σ sensitivity. Try reparametrized variant (Dănăilă & Buiu 2024). If still unstable, the problem is deeper than the loss function — escalate to Path C (different generative model).

### Gate 2: After Path E (#37)

**Question:** Is long-horizon CRPS acceptable with teacher forcing?

- **If yes:** Evaluate whether Paths D and B are needed (diagnostic-gated).
- **If no with scheduled sampling:** Upgrade to GTF (adaptive α, Hess et al. 2023).
- **If no even with GTF:** The autoregressive architecture itself is the bottleneck. Escalate to Path F.

### Gate 3: After Paths D/B (#38, #39)

**Question:** Is overall forecasting quality sufficient for operational deployment?

- **If yes:** Done. Ship the Tobit + teacher forcing + tail/selection corrections model.
- **If no:** The Tobit framework has hit a ceiling. Pursue Paths C and/or F as strategic architecture explorations.

---

## Literature Foundation

This roadmap is grounded in 10 papers from 7 independent research groups:

| ID | Paper | Core contribution to this roadmap |
|----|-------|----------------------------------|
| P1 | [Dănăilă & Buiu 2024](https://link.springer.com/article/10.1007/s10044-024-01216-9) | Tobit loss for DNNs; globally concave reparametrized variant |
| P2 | [Jacobson & Zou 2022](https://arxiv.org/abs/2203.02601) | Tobit MSE advantage grows with censoring proportion |
| P3 | [Jacobson & Zou 2023](https://ideas.repec.org/a/taf/jnlbes/v42y2024i1p286-297.html) | Penalized Tobit with oracle property; HIV viral load application |
| P4 | [Zhang et al. 2021](https://www.sciencedirect.com/science/article/abs/pii/S0893608021003531) | ReLU ≡ Tobit-I; DTN-II Heckman selection with deep networks |
| P5 | [Wu et al. 2026](https://link.springer.com/article/10.1007/s10985-026-09690-5) | Deep Tobit with variable selection; convergence guarantees |
| P6 | [O'Neill 2024](https://arxiv.org/abs/2211.07506) | Tobit works with nonlinear models (BART); nonparametric errors |
| P7 | [Iacus et al. 2025](https://arxiv.org/abs/2512.21435) | ZINB on VIEWS/PRIO-Grid data; direct MH; R²=0.955 |
| P8 | Wilson et al. KDD 2022 | Hurdle underpredicts extremes; three-component mixture |
| P9 | [Hess et al. ICML 2023](https://proceedings.mlr.press/v202/hess23a.html) | Adaptive teacher forcing; Jacobian-based α; bounded gradients |
| P10 | [Bengio et al. NeurIPS 2015](https://papers.nips.cc/paper/2015/hash/e995f98d56967d946471af29d7bf99f1-Abstract.html) | Scheduled sampling curriculum |

---

## Current Status

- [x] Root cause diagnosed (hurdle mask gradient starvation)
- [x] Isolation plan designed (`reports/sweep_isolation_plan.md`)
- [x] S0 baseline confirmed stable
- [x] S1 target_weights confirmed safe
- [ ] S2 hurdle result pending (expected FAIL — training dynamics show non-convergence)
- [ ] S3 qs99 pending
- [ ] S5 basu_dpd pending
- [ ] Literature survey complete (`reports/paths_forward.md`)
- [ ] **Path A implementation** ← NEXT

**Comment by Polichinel** [comment]
## Gate 1 Status Update (2026-05-29)

**Gate 1 PASSES — Path A validated. Proceeding to Path E.**

S2a-150 experiment (Tobit, sigma=1.0, 150 lessons) confirms:
- CRPS 43-80% better than S0 baseline across all targets
- MCR recovers from severe underprediction (0.04→0.56 for sb, 0.003→0.81 for ns) with more training
- Training converges by ~lesson 60; 150 lessons adds no value beyond ~80
- Exposure bias quantified: step-wise MCR gap (0.56 vs month-wise 0.98) provides Gate 2 baseline

**Residual concerns (registered in risk register):**
- C-95: MCR asymmetry (lr_sb=0.56 overshoot vs lr_os=0.015 undershoot)
- C-96: Tobit converges faster than MSE — total_lessons should be 60-80, not 150
- C-97: Step-wise CRPS degradation at long horizons (13% for sb) — exposure bias baseline for Path E
- C-98: Implicit input_channels == 3 × output_channels invariant needs validation before Path E
- C-99: reg_latent vs reg dual-path hazard for scheduled sampling

**Updated checklist:**
- [x] Root cause diagnosed (hurdle mask gradient starvation)
- [x] Isolation plan designed
- [x] S0 baseline confirmed stable
- [x] S1 target_weights confirmed safe
- [x] S2 hurdle FAIL confirmed
- [x] S2a Tobit PASS (20 lessons)
- [x] S2a-150 Tobit PASS (150 lessons, MCR recovery confirmed)
- [x] **Gate 1 PASS — proceed to Path E**
- [ ] Path E implementation (issue #37) ← NEXT

**Comment by Polichinel** [comment]
## Gate 2 Status Update (2026-06-01)

**Gate 2 PASSES — Path E validated. Scheduled sampling eliminates exposure bias.**

The scheduled sampling sweep (eps_max ∈ {0.0, 0.25, 0.5, 0.75}, linear schedule, warmup=10, 80 lessons) confirms:

| eps_max | sb step-MCR | month/step gap | sb CRPS |
|---------|-------------|----------------|---------|
| 0.00 (control) | 1.92 | **1.60** | 0.265 |
| 0.50 (recommended) | 0.37 | **0.02** | 0.152 |

- Exposure bias gap: 1.60 → 0.02 (**99% reduction**)
- Step-wise sb CRPS: 0.265 → 0.152 (**43% improvement**)
- No escalation to GTF needed

**Updated checklist:**
- [x] Root cause diagnosed (hurdle mask gradient starvation)
- [x] S0 baseline confirmed stable
- [x] S1 target_weights confirmed safe
- [x] S2 hurdle FAIL confirmed
- [x] S2a Tobit PASS (Gate 1)
- [x] Sigma sensitivity sweep → per-target sigma (issue #44, PR #47)
- [x] Learnable sigma validated (ADR-055)
- [x] **Gate 1 PASS — Tobit restores convergence**
- [x] Scheduled sampling implemented (issue #37, PR #50, ADR-056)
- [x] **Gate 2 PASS — exposure bias eliminated**
- [ ] Gate 3: evaluate whether Paths D/B are needed (diagnostic-gated)

**Literature:**
- Bengio, S. et al. (2015). "Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks." NeurIPS 2015.
- Hess, F. et al. (2023). "Generalized Teacher Forcing for Learning Chaotic Dynamics." ICML 2023. (escalation path — not needed)

**Recommended production config:**
```python
'loss_reg': 'tobit',
'loss_reg_sigma': {'lr_sb_best': 1.0, 'lr_ns_best': 0.75, 'lr_os_best': 0.5},
'ss_schedule': 'linear',
'ss_warmup_lessons': 10,
'ss_epsilon_max': 0.5,
'total_lessons': 80,
```

**Comment by Polichinel** [comment]
**PARKED** — deferred fallback (not rejected); see ZINB epic #97 / checklist #104 / C-139. Revisit only if the ZINB head fails. Rationale: loss/calibration investigate-X — deferred; ZINB is the chosen direction
